[CSS URL Request Modifiers] Add parsing and serialization support for the CSS URL request modifier functions

weinig opened the pull request:
#### f77f35baa3ee1c79de24e39ed63a2c2170bdac20
<pre>
[CSS URL Request Modifiers] Add parsing and serialization support for the CSS URL request modifier functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=291271">https://bugs.webkit.org/show_bug.cgi?id=291271</a>

Reviewed by Anne van Kesteren and Antti Koivisto.

Adds support for the CSS Values and Units 5 URL request modifier functions,
crossorigin(), integrity() and referrerpolicy() (<a href="https://drafts.csswg.org/css-values-5/#request-url-modifiers).">https://drafts.csswg.org/css-values-5/#request-url-modifiers).</a>

The new functionality is off by default behind the experimental feature
settings flag: CSSURLModifiersEnabled.

Still needed is support applying the modifiers to style resource requests.

In addition to storing the new modifiers, the LoadedFromOpaqueSource bit which
was being passed manually is now included in the new modifiers struct.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/StyleRuleImport.cpp:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
* Source/WebCore/css/values/primitives/CSSURL.cpp:
* Source/WebCore/css/values/primitives/CSSURL.h:
* Source/WebCore/css/values/primitives/CSSURLModifiers.cpp: Added.
* Source/WebCore/css/values/primitives/CSSURLModifiers.h: Added.
* Source/WebCore/html/HTMLBodyElement.cpp:
* Source/WebCore/html/HTMLLinkElement.cpp:
* Source/WebCore/html/HTMLTableElement.cpp:
* Source/WebCore/html/HTMLTablePartElement.cpp:
* Source/WebCore/loader/LoadedFromOpaqueSource.h: Added.
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/cache/CachedSVGDocumentReference.cpp:
* Source/WebCore/loader/cache/CachedSVGDocumentReference.h:
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/style/values/filter-effects/StyleFilterReference.cpp:
* Source/WebCore/style/values/primitives/StyleURL.cpp:
* Source/WebCore/style/values/primitives/StyleURL.h:
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:

Canonical link: <a href="https://commits.webkit.org/293495@main">https://commits.webkit.org/293495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b65e52275b6094554efe1a6262704d7a1d0559ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99095 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104215 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/49676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101135 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/19027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27175 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/75427 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/49676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102100 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/19027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89483 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/55791 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/19027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/7455 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49052 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/91774 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/19027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7543 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106578 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/97714 "Built successfully and passed tests") | [❌ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26204 "Failed to checkout and rebase branch from PR 43802") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/19098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84384 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26567 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85683 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/83896 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21274 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/28560 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/6236 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/19917 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26145 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/31326 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/121329 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/25966 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/33913 "Found 9 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/es6/ES6MathAPIs.js.default, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.default-wasm, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-bbq, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-bbq-no-consts, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-collect-continuously, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-eager, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-eager-jettison, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-no-cjit, wasm.yaml/wasm/stress/i32_trunc_sat_f32_s.js.wasm-slow-memory (failure)") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/29279 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27532 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk reviewed (approved):
Tests look reasonable, though we should add actual networking tests as well (prolly ref tests) before enabling. Loader logic looks good as well to the extent I'm familiar with it. But please don't consider this approval for the CSS changes. I'm far less familiar with that code.

weinig committed:
[CSS URL Request Modifiers] Add parsing and serialization support for the CSS URL request modifier functions
https://bugs.webkit.org/show_bug.cgi?id=291271

Reviewed by Anne van Kesteren and Antti Koivisto.

Adds support for the CSS Values and Units 5 URL request modifier functions,
crossorigin(), integrity() and referrerpolicy() (https://drafts.csswg.org/css-values-5/#request-url-modifiers).

The new functionality is off by default behind the experimental feature
settings flag: CSSURLModifiersEnabled.

Still needed is support applying the modifiers to style resource requests.

In addition to storing the new modifiers, the LoadedFromOpaqueSource bit which
was being passed manually is now included in the new modifiers struct.

* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub.html: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-computed.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-invalid.sub-expected.txt: Added.
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-values/urls/url-request-modifiers-serialize.sub-expected.txt: Added.
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/StyleRuleImport.cpp:
* Source/WebCore/css/StyleSheetContents.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UI.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+URL.cpp:
* Source/WebCore/css/values/primitives/CSSURL.cpp:
* Source/WebCore/css/values/primitives/CSSURL.h:
* Source/WebCore/css/values/primitives/CSSURLModifiers.cpp: Added.
* Source/WebCore/css/values/primitives/CSSURLModifiers.h: Added.
* Source/WebCore/html/HTMLBodyElement.cpp:
* Source/WebCore/html/HTMLLinkElement.cpp:
* Source/WebCore/html/HTMLTableElement.cpp:
* Source/WebCore/html/HTMLTablePartElement.cpp:
* Source/WebCore/loader/LoadedFromOpaqueSource.h: Added.
* Source/WebCore/loader/ResourceLoaderOptions.h:
* Source/WebCore/loader/cache/CachedSVGDocumentReference.cpp:
* Source/WebCore/loader/cache/CachedSVGDocumentReference.h:
* Source/WebCore/rendering/style/ReferenceFilterOperation.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.cpp:
* Source/WebCore/rendering/style/StyleCursorImage.h:
* Source/WebCore/style/values/filter-effects/StyleFilterReference.cpp:
* Source/WebCore/style/values/primitives/StyleURL.cpp:
* Source/WebCore/style/values/primitives/StyleURL.h:
* Source/WebCore/svg/SVGFontFaceUriElement.cpp:

Canonical link: https://commits.webkit.org/293495@main
